File: csharp/MudbaseShowcase.Kanban/Services/ActivityText.cs
using MudbaseShowcase.Kanban.Models;

namespace MudbaseShowcase.Kanban.Services;

/// <summary>Renders one activity row as a plain, readable sentence for the feed - direct port of web/src/lib/activity-text.ts's describeActivity.</summary>
public static class ActivityText
{
    public static string Describe(ActivityDocument entry)
    {
        string who = string.IsNullOrWhiteSpace(entry.ActorName) ? "Someone" : entry.ActorName;

        return entry.Action switch
        {
            ActivityActions.CreatedCard =>
                $"{who} created card \"{entry.CardTitle ?? "Untitled"}\" in {entry.ToList ?? "a list"}",

            ActivityActions.Moved when entry.FromList is not null && entry.ToList is not null && entry.FromList == entry.ToList =>
                $"{who} reordered \"{entry.CardTitle ?? "a card"}\" within {entry.ToList}",

            ActivityActions.Moved =>
                $"{who} moved \"{entry.CardTitle ?? "a card"}\" from {entry.FromList ?? "?"} to {entry.ToList ?? "?"}",

            ActivityActions.DeletedCard =>
                $"{who} deleted card \"{entry.CardTitle ?? "Untitled"}\" from {entry.FromList ?? "a list"}",

            ActivityActions.CreatedList =>
                $"{who} created list \"{entry.ToList ?? "Untitled"}\"",

            ActivityActions.RenamedList =>
                $"{who} renamed list \"{entry.FromList ?? "?"}\" to \"{entry.ToList ?? "?"}\"",

            ActivityActions.DeletedList =>
                $"{who} deleted list \"{entry.FromList ?? "Untitled"}\"",

            _ => $"{who} did something on the board",
        };
    }
}

File: php/src/Support/ActivityText.php
<?php

declare(strict_types=1);

namespace App\Support;

/**
 * Renders one activity row as a plain, readable sentence for the feed. A line-for-line PHP port
 * of the reference app's `describeActivity` (src/lib/activity-text.ts).
 */
final class ActivityText
{
    /** @param array<string, mixed> $entry */
    public static function describe(array $entry): string
    {
        $who = is_string($entry['actorName'] ?? null) && $entry['actorName'] !== '' ? $entry['actorName'] : 'Someone';
        $action = is_string($entry['action'] ?? null) ? $entry['action'] : '';
        $cardTitle = is_string($entry['cardTitle'] ?? null) ? $entry['cardTitle'] : null;
        $fromList = is_string($entry['fromList'] ?? null) ? $entry['fromList'] : null;
        $toList = is_string($entry['toList'] ?? null) ? $entry['toList'] : null;

        return match ($action) {
            'created_card' => "{$who} created card \"" . ($cardTitle ?? 'Untitled') . '" in ' . ($toList ?? 'a list'),
            'moved' => ($fromList !== null && $toList !== null && $fromList === $toList)
                ? "{$who} reordered \"" . ($cardTitle ?? 'a card') . "\" within {$toList}"
                : "{$who} moved \"" . ($cardTitle ?? 'a card') . '" from ' . ($fromList ?? '?') . ' to ' . ($toList ?? '?'),
            'deleted_card' => "{$who} deleted card \"" . ($cardTitle ?? 'Untitled') . '" from ' . ($fromList ?? 'a list'),
            'created_list' => "{$who} created list \"" . ($toList ?? 'Untitled') . '"',
            'renamed_list' => "{$who} renamed list \"" . ($fromList ?? '?') . '" to "' . ($toList ?? '?') . '"',
            'deleted_list' => "{$who} deleted list \"" . ($fromList ?? 'Untitled') . '"',
            default => "{$who} did something on the board",
        };
    }
}
